### PR TITLE
Revert "Add support fchstat and chstat  function for littlefs"

### PR DESCRIPTION
**Pull request by yamt**
Reverts apache/nuttx#13964

because:
* it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.
* we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.
* it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
can we add an option to make the new feature selectable?

**Comment by yamt** [comment]
> can we add an option to make the new feature selectable?

it's possible. but it's better to remove it for now, IMO.
adding more and more local patches like this here makes the use of different versions of littlefs impossible.

File: fs/littlefs/lfs_vfs.c [hunk 1]
```diff
@@ -1572,15 +1571,6 @@ static int littlefs_stat(FAR struct inode *mountpt, FAR const char *relpath,
   buf->st_blocks       = (buf->st_size + buf->st_blksize - 1) /
                          buf->st_blksize;
 
-  if (info.type == LFS_TYPE_REG)
```
Comment (xiaoxiang781216) [inline fs/littlefs/lfs_vfs.c:1683]: need keep this patch

File: fs/littlefs/lfs_vfs.c [hunk 2]
```diff
@@ -1575,7 +1556,6 @@ static int littlefs_stat(FAR struct inode *mountpt, FAR const char *relpath,
 
       ret = 0;
       memset(&attr, 0, sizeof(attr));
-      attr.at_mode = S_IRWXG | S_IRWXU | S_IRWXO;
```
Comment (xiaoxiang781216) [inline fs/littlefs/lfs_vfs.c:1666]: need keep this change

File: fs/littlefs/lfs_vfs.c [hunk 3]
```diff
@@ -1573,11 +1573,11 @@ static int littlefs_stat(FAR struct inode *mountpt, FAR const char *relpath,
           goto errout;
         }
 
+      ret = 0;
```
Comment (xiaoxiang781216) [inline fs/littlefs/lfs_vfs.c]: need keep this line change

**Comment by yamt** [comment]
@xiaoxiang781216 
i'm not sure which changes you are talking about. can you explain a bit?

anyway, i guess the original PR should not have contained unrelated changes together.
i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> 
> anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.

Ok, let's me revert.

**Comment by yamt** [comment]
> > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> 
> Ok, let's me revert.

what's the status of this?

**Comment by xiaoxiang781216** [comment]
I will send in this weak.

**Comment by xiaoxiang781216** [comment]
> > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> 
> Ok, let's me revert.

@yamt After comparing carefully all nuttx and littlefs patch, I think the change is good in general. The follow is my answer to your concern.

> Reverts #13964
> 
> because:
> 
> * it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.

Could you give a repro step? So, we can provide a fix.

> * we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.

It upstream here: https://github.com/littlefs-project/littlefs/pull/1045.
littlefs already provide lfs_setattr and lfs_getattr which work on file path, the new functions (lfs_file_getattr and lfs_file_setattr) which work on file handle is a natural extension. Let's wait the author feedback.

> * it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.

uid/gid is useful after:
https://github.com/apache/nuttx/pull/8924
https://github.com/apache/nuttx/pull/10119
https://github.com/apache/nuttx/pull/10176
https://github.com/apache/nuttx-apps/pull/1691
This is the first real filesystem supported uid/gid/mode on NuttX, so it's better to keep it to demonstrate this capability.

BTW, the attribute contains the date/time information which is very useful for most people.

**Comment by yamt** [comment]
> > > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> > 
> > 
> > Ok, let's me revert.
> 
> @yamt After comparing carefully all nuttx and littlefs patch, I think the change is good in general. The follow is my answer to your concern.
> 
> > Reverts #13964
> > because:
> > 
> > * it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.
> 
> Could you give a repro step? So, we can provide a fix.

you can reproduce it with the instructions in https://nuttx.apache.org/docs/latest/platforms/xtensa/esp32s3/boards/esp32s3-devkit/index.html#toywasm.
(with an update https://github.com/apache/nuttx/pull/14832)

> > * we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.
> 
> It upstream here: [littlefs-project/littlefs#1045](https://github.com/littlefs-project/littlefs/pull/1045). littlefs already provide lfs_setattr and lfs_getattr which work on file path, the new functions (lfs_file_getattr and lfs_file_setattr) which work on file handle is a natural extension. Let's wait the author feedback.
> 
> > * it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.
> 
> uid/gid is useful after: #8924 #10119 #10176 [apache/nuttx-apps#1691](https://github.com/apache/nuttx-apps/pull/1691) This is the first real filesystem supported uid/gid/mode on NuttX, so it's better to keep it to demonstrate this capability.
> 
> BTW, the attribute contains the date/time information which is very useful for most people.

i don't object much if:
* the patch is accepted by the upstream. (it's important for me to be able to use latest version of littlefs.)
* and it's optional. (i don't want to use my flash space for these attributes.)

**Comment by xiaoxiang781216** [comment]
> > > > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > > > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> > > 
> > > 
> > > Ok, let's me revert.
> > 
> > 
> > @yamt After comparing carefully all nuttx and littlefs patch, I think the change is good in general. The follow is my answer to your concern.
> > > Reverts #13964
> > > because:
> > > 
> > > * it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.
> > 
> > 
> > Could you give a repro step? So, we can provide a fix.
> 
> you can reproduce it with the instructions in https://nuttx.apache.org/docs/latest/platforms/xtensa/esp32s3/boards/esp32s3-devkit/index.html#toywasm. (with an update #14832)
> 

@crafcat7 please find the root cause why the format fail with attribute patch.

> > > * we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.
> > 
> > 
> > It upstream here: [littlefs-project/littlefs#1045](https://github.com/littlefs-project/littlefs/pull/1045). littlefs already provide lfs_setattr and lfs_getattr which work on file path, the new functions (lfs_file_getattr and lfs_file_setattr) which work on file handle is a natural extension. Let's wait the author feedback.
> > > * it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.
> > 
> > 
> > uid/gid is useful after: #8924 #10119 #10176 [apache/nuttx-apps#1691](https://github.com/apache/nuttx-apps/pull/1691) This is the first real filesystem supported uid/gid/mode on NuttX, so it's better to keep it to demonstrate this capability.
> > BTW, the attribute contains the date/time information which is very useful for most people.
> 
> i don't object much if:
> 
> * the patch is accepted by the upstream. (it's important for me to be able to use latest version of littlefs.)

the patch already upstream, let's wait the feedback.

> * and it's optional. (i don't want to use my flash space for these attributes.)

Ok, @crafcat7 please add an option to skip the attribute operation.

**Comment by crafcat7** [comment]
> > > > > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > > > > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> > > > 
> > > > 
> > > > Ok, let's me revert.
> > > 
> > > 
> > > @yamt After comparing carefully all nuttx and littlefs patch, I think the change is good in general. The follow is my answer to your concern.
> > > > Reverts #13964
> > > > because:
> > > > 
> > > > * it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.
> > > 
> > > 
> > > Could you give a repro step? So, we can provide a fix.
> > 
> > 
> > you can reproduce it with the instructions in https://nuttx.apache.org/docs/latest/platforms/xtensa/esp32s3/boards/esp32s3-devkit/index.html#toywasm. (with an update #14832)
> 
> @crafcat7 please find the root cause why the format fail with attribute patch.

The reason for the fstat failure should be caused during mkfs.
Because the get / set_attr function is introduced in the littlefs currently used in NuttX, we need to write the attributes of the original data into the actual mkfs through lfs_setattr.
Based on https://github.com/whitecatboard/Lua-RTOS-ESP32/blob/master/components/mklfs/src/mklfs.c
It can be implemented as follows
```
static void create_file(char *src) {
    char *path;
    int ret;

    path = strchr(src, '/');
    if (path) {
        fprintf(stdout, "%s\r\n", path);

        // Open source file
        FILE *srcf = fopen(src,"rb");
        if (!srcf) {
            fprintf(stderr,"can't open source file %s: errno=%d (%s)\r\n", src, errno, strerror(errno));
            exit(1);
        }

        // Get source file fd
        int fd = fileno(srcf);

        // Get file metadata
        struct stat filestat;
        if (fstat(fd, &filestat) == -1) {
            fprintf(stderr,"can't get source file %s: errno=%d (%s)\r\n", src, errno, strerror(errno));
            fclose(srcf);
            exit(1);
        }

        // Open destination file
        lfs_file_t dstf;
        if ((ret = lfs_file_open(&lfs, &dstf, path, LFS_O_WRONLY | LFS_O_CREAT)) < 0) {
            fprintf(stderr,"can't open destination file %s: error=%d\r\n", path, ret);
            exit(1);
        }

		char c = fgetc(srcf);
		while (!feof(srcf)) {
			ret = lfs_file_write(&lfs, &dstf, &c, 1);
			if (ret < 0) {
				fprintf(stderr,"can't write to destination file %s: error=%d\r\n", path, ret);
				exit(1);
			}
			c = fgetc(srcf);
		}

    // Attributes of written data
    struct lfs_stat attr;
    memset(&attr, 0, sizeof(attr));

    attr.at_atim = filestat.st_atime;
    attr.at_mtim = filestat.st_mtime;
    attr.at_ctim = filestat.st_ctime;
    attr.at_gid  = filestat.st_gid;
    attr.at_uid  = filestat.st_uid;
    attr.at_mode = filestat.st_mode;
    attr.at_size = filestat.st_size;

    ret = lfs_setattr(&lfs, path, 0, &attr, sizeof(attr));
    if (ret < 0) {
      fprintf(stderr,"can't set attr to destination file %s: error=%d\r\n", path, ret);
      exit(1);
    }

    // Close destination file
		ret = lfs_file_close(&lfs, &dstf);
		if (ret < 0) {
			fprintf(stderr,"can't close destination file %s: error=%d\r\n", path, ret);
			exit(1);
		}

        // Close source file
        fclose(srcf);
    }
}
```
Because the get / set_attr function is introduced in the littlefs currently used in NuttX, in the actual mkfs, we need to write the attributes of the original data into it through lfs_setattr.
Based on https://github.com/whitecatboard/Lua-RTOS-ESP32/blob/master/components/mklfs/src/mklfs.c
It can be implemented as follows
At the same time, I will submit a compatibility change at the party time. When lfs_file_getattr returns LFS_ERR_NOENT, just return 0 (ignore it), so there will be no problem. The file size is obtained through `lfs_file_size`, so the file size is not affected


> > > > * we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.
> > > 
> > > 
> > > It upstream here: [littlefs-project/littlefs#1045](https://github.com/littlefs-project/littlefs/pull/1045). littlefs already provide lfs_setattr and lfs_getattr which work on file path, the new functions (lfs_file_getattr and lfs_file_setattr) which work on file handle is a natural extension. Let's wait the author feedback.
> > > > * it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.
> > > 
> > > 
> > > uid/gid is useful after: #8924 #10119 #10176 [apache/nuttx-apps#1691](https://github.com/apache/nuttx-apps/pull/1691) This is the first real filesystem supported uid/gid/mode on NuttX, so it's better to keep it to demonstrate this capability.
> > > BTW, the attribute contains the date/time information which is very useful for most people.
> > 
> > 
> > i don't object much if:
> > 
> > * the patch is accepted by the upstream. (it's important for me to be able to use latest version of littlefs.)
> 
> the patch already upstream, let's wait the feedback.
> 
> > * and it's optional. (i don't want to use my flash space for these attributes.)
> 
> Ok, @crafcat7 please add an option to skip the attribute operation.

Likewise, I will submit a change later to turn off the property setting.

**Comment by xiaoxiang781216** [comment]
@crafcat7 let's port https://github.com/whitecatboard/Lua-RTOS-ESP32/blob/master/components/mklfs/src/mklfs.c to apps/fs/utils and add a command line option to generate the compatible attr entry we expect.

**Comment by crafcat7** [comment]
> @crafcat7 let's port https://github.com/whitecatboard/Lua-RTOS-ESP32/blob/master/components/mklfs/src/mklfs.c to apps/fsutils and add a command line option to generate the compatible attr entry we expect.

I will then submit a PR under apps to complete it

**Comment by crafcat7** [comment]
> > > > > > @xiaoxiang781216 i'm not sure which changes you are talking about. can you explain a bit?
> > > > > > anyway, i guess the original PR should not have contained unrelated changes together. i suspect it's simpler to revert the PR as a whole and ask the author to re-submit those parts separately.
> > > > > 
> > > > > 
> > > > > Ok, let's me revert.
> > > > 
> > > > 
> > > > @yamt After comparing carefully all nuttx and littlefs patch, I think the change is good in general. The follow is my answer to your concern.
> > > > > Reverts #13964
> > > > > because:
> > > > > 
> > > > > * it seems to introduce some regressions. i saw fstat failing with ENOENT on files on an image generated with https://github.com/jrast/littlefs-python.
> > > > 
> > > > 
> > > > Could you give a repro step? So, we can provide a fix.
> > > 
> > > 
> > > you can reproduce it with the instructions in https://nuttx.apache.org/docs/latest/platforms/xtensa/esp32s3/boards/esp32s3-devkit/index.html#toywasm. (with an update #14832)
> > 
> > 
> > @crafcat7 please find the root cause why the format fail with attribute patch.
> 
> The reason for the fstat failure should be caused during mkfs. Because the get / set_attr function is introduced in the littlefs currently used in NuttX, we need to write the attributes of the original data into the actual mkfs through lfs_setattr. Based on https://github.com/whitecatboard/Lua-RTOS-ESP32/blob/master/components/mklfs/src/mklfs.c It can be implemented as follows
> 
> ```
> static void create_file(char *src) {
>     char *path;
>     int ret;
> 
>     path = strchr(src, '/');
>     if (path) {
>         fprintf(stdout, "%s\r\n", path);
> 
>         // Open source file
>         FILE *srcf = fopen(src,"rb");
>         if (!srcf) {
>             fprintf(stderr,"can't open source file %s: errno=%d (%s)\r\n", src, errno, strerror(errno));
>             exit(1);
>         }
> 
>         // Get source file fd
>         int fd = fileno(srcf);
> 
>         // Get file metadata
>         struct stat filestat;
>         if (fstat(fd, &filestat) == -1) {
>             fprintf(stderr,"can't get source file %s: errno=%d (%s)\r\n", src, errno, strerror(errno));
>             fclose(srcf);
>             exit(1);
>         }
> 
>         // Open destination file
>         lfs_file_t dstf;
>         if ((ret = lfs_file_open(&lfs, &dstf, path, LFS_O_WRONLY | LFS_O_CREAT)) < 0) {
>             fprintf(stderr,"can't open destination file %s: error=%d\r\n", path, ret);
>             exit(1);
>         }
> 
> 		char c = fgetc(srcf);
> 		while (!feof(srcf)) {
> 			ret = lfs_file_write(&lfs, &dstf, &c, 1);
> 			if (ret < 0) {
> 				fprintf(stderr,"can't write to destination file %s: error=%d\r\n", path, ret);
> 				exit(1);
> 			}
> 			c = fgetc(srcf);
> 		}
> 
>     // Attributes of written data
>     struct lfs_stat attr;
>     memset(&attr, 0, sizeof(attr));
> 
>     attr.at_atim = filestat.st_atime;
>     attr.at_mtim = filestat.st_mtime;
>     attr.at_ctim = filestat.st_ctime;
>     attr.at_gid  = filestat.st_gid;
>     attr.at_uid  = filestat.st_uid;
>     attr.at_mode = filestat.st_mode;
>     attr.at_size = filestat.st_size;
> 
>     ret = lfs_setattr(&lfs, path, 0, &attr, sizeof(attr));
>     if (ret < 0) {
>       fprintf(stderr,"can't set attr to destination file %s: error=%d\r\n", path, ret);
>       exit(1);
>     }
> 
>     // Close destination file
> 		ret = lfs_file_close(&lfs, &dstf);
> 		if (ret < 0) {
> 			fprintf(stderr,"can't close destination file %s: error=%d\r\n", path, ret);
> 			exit(1);
> 		}
> 
>         // Close source file
>         fclose(srcf);
>     }
> }
> ```
> 
> And I will submit a compatibility change that is when lfs_file_getattr returns LFS_ERR_NOENT, just return 0 (ignore it), so there will be no problem. The file size is obtained through `lfs_file_size`, so the file size is not affected
> 
> > > > > * we should reduce the amount of the local patches. not the opposite. if you want to extend littlefs, please upstream it first.
> > > > 
> > > > 
> > > > It upstream here: [littlefs-project/littlefs#1045](https://github.com/littlefs-project/littlefs/pull/1045). littlefs already provide lfs_setattr and lfs_getattr which work on file path, the new functions (lfs_file_getattr and lfs_file_setattr) which work on file handle is a natural extension. Let's wait the author feedback.
> > > > > * it's controversial if it's a good idea to waste on-disk structures for attributes like file modes and uid/gid, which nuttx doesn't really support.
> > > > 
> > > > 
> > > > uid/gid is useful after: #8924 #10119 #10176 [apache/nuttx-apps#1691](https://github.com/apache/nuttx-apps/pull/1691) This is the first real filesystem supported uid/gid/mode on NuttX, so it's better to keep it to demonstrate this capability.
> > > > BTW, the attribute contains the date/time information which is very useful for most people.
> > > 
> > > 
> > > i don't object much if:
> > > 
> > > * the patch is accepted by the upstream. (it's important for me to be able to use latest version of littlefs.)
> > 
> > 
> > the patch already upstream, let's wait the feedback.
> > > * and it's optional. (i don't want to use my flash space for these attributes.)
> > 
> > 
> > Ok, @crafcat7 please add an option to skip the attribute operation.
> 
> Likewise, I will submit a change later to turn off the property setting.

I submitted this compatibility fix at https://github.com/apache/nuttx/pull/14844

**Comment by xiaoxiang781216** [comment]
@yamt could you abandon this pr since https://github.com/apache/nuttx/pull/14844 get merged.

**Comment by yamt** [comment]
> @yamt could you abandon this pr since #14844 get merged.

as your PR (https://github.com/littlefs-project/littlefs/pull/1045) doesn't seem being accepted by the upstream, i guess we should still merge this revert.

**Comment by crafcat7** [comment]
I submitted a change to replace lfs_file_xxxattr with the existing method in littlefs
Until the littlefs community accepts lfs_file_xxxattr, we will continue to use the changes at this PR: https://github.com/apache/nuttx/pull/14901

**Comment by xiaoxiang781216** [comment]
@yamt could we close this pr now?

**Comment by yamt** [comment]
> @yamt could we close this pr now?

sure.

**Pull request closed without merging**